### PR TITLE
added @covers annotation to test classes

### PR DESCRIPTION
**Pull request by OskarStark**
Does this make sense?

### CODE REVIEW

**Comment by greg0ire** [comment]
Don't do one commit per file, you're doing the same thing (or at least your commit messages show you do)

File: Tests/Admin/AdminHelperTest.php [hunk 1]
```diff
@@ -18,6 +18,9 @@
 
 class AdminHelperTest extends \PHPUnit_Framework_TestCase
 {
+    /**
+     * @covers Sonata\AdminBundle\Admin\AdminHelper::getChildFormBuilder
+     */
```
Comment (greg0ire) [inline Tests/Admin/AdminHelperTest.php]: Not sure it makes sense. I think this annotation is for when you are not obviously covering a method. The name of your test method and the fact that you call the method directly makes it obvious.

Comment (OskarStark) [inline Tests/Admin/AdminHelperTest.php]: the background of this change is, that i would like to know which parts are not directly covered by a test. Creating an Admin in a test does not mean it has a test and i want to find out which parts needs to be tested "better" otherwise code coverage would say, yes this is covered!

anyway, we currently have some `@covers` annotations in our project, for example in `AdminTest`

Comment (greg0ire) [inline Tests/Admin/AdminHelperTest.php]: Oh ok, so `@covers` makes sure the coverage doesn't say this method covers something else? But what if the indirectly covered method is a protected or private method ?

Comment (OskarStark) [inline Tests/Admin/AdminHelperTest.php]: > Oh ok, so @covers makes sure the coverage doesn't say this method covers something else?

exactly 👍  so we could get a nice overview where we need to invest time/work/code. Currently i have a collegue who wants to write tests and asked me what he should test ;-)

> But what if the indirectly covered method is a protected or private method ?

seems to be no problem

Comment (OskarStark) [inline Tests/Admin/AdminHelperTest.php]: @core23 what do you think?

Comment (OskarStark) [inline Tests/Admin/AdminHelperTest.php]: you can find more information here @greg0ire 

https://phpunit.de/manual/current/en/code-coverage-analysis.html

File: Tests/Admin/AdminTest.php [hunk 2]
```diff
@@ -482,6 +511,9 @@ public function testChildren()
 
     /**
      * @covers Sonata\AdminBundle\Admin\Admin::configure
+     * @covers Sonata\AdminBundle\Admin\Admin::getUniqid
+     * @covers Sonata\AdminBundle\Admin\Admin::getClassnameLabel
+     * @covers Sonata\AdminBundle\Admin\Admin::setClassnameLabel
```
Comment (OskarStark) [inline Tests/Admin/AdminTest.php]: wtf?

This method does not exist?! @greg0ire 

```
PHPUnit 4.8.23 by Sebastian Bergmann and contributors.

...................Trying to @cover or @use not existing method "Sonata\AdminBundle\Admin\Admin::setClassnameLabel".
```

Comment (OskarStark) [inline Tests/Admin/AdminTest.php]: ah ok this are methods in fixture files 😄

File: Tests/Admin/AdminHelperTest.php (hunk 1 above)
Comment (Soullivaneuh) [inline Tests/Admin/AdminHelperTest.php]: I'm not a big fan of `@covers` on method.

AFAIK, is you write `@covers Some::thing` and `Some::that` is called, it would not be covered even if it's tested.

Maybe `@covers` on the class?

Comment (OskarStark) [inline Tests/Admin/AdminHelperTest.php]: > AFAIK, is you write @covers Some::thing and Some::that is called, it would not be covered even if it's tested.

That is exactly the point, calling the constructor for example, does not mean it is tested/covered. If you'd say this is covered you need to explicitly add `@covers \AppBundel\Model\Foo:__construct`

Otherwise we have green lines, but that only means they are executed, not covered

Comment (Soullivaneuh) [inline Tests/Admin/AdminHelperTest.php]: Yes, but what do you do if you assert a result from multiple method calls (e.g. private ones)?

Comment (greg0ire) [inline Tests/Admin/AdminHelperTest.php]: I guess you use multiple @covers annotations in that case.

Comment (Soullivaneuh) [inline Tests/Admin/AdminHelperTest.php]: > I guess you use multiple @covers annotations in that case.

Yes if you don't forgot some. This is why I'm not a big fan human error with that are too risky IMHO.

This is my vote, but not a veto of course.

Comment (greg0ire) [inline Tests/Admin/AdminHelperTest.php]: > too risky

The risk being what ? Having less code coverage reported ? Not a big risk IMO

But I think we should not use `@covers` for every method, only for public methods that call other public methods.

Comment (Soullivaneuh) [inline Tests/Admin/AdminHelperTest.php]: > But I think we should not use @covers for every method, only for public methods that call other public methods.

Ok but we are going back to the same problem: What for testing private/protected methods?

Comment (greg0ire) [inline Tests/Admin/AdminHelperTest.php]: Same solution. Here is what we should do IMO

```php
<?php
/**
 * @covers a
 * @covers c
 */
public function a ()
{
   $this->b();
   $this->c();
}
public function b ()
{
}
private function c ()
{
}
```

Comment (Soullivaneuh) [inline Tests/Admin/AdminHelperTest.php]: I'm confused, you just said:

> But I think we should not use @covers for every method, only for public methods that call other public methods.

Maybe I misunderstood?

Comment (greg0ire) [inline Tests/Admin/AdminHelperTest.php]: Well, doesn't a call b ?

Comment (OskarStark) [inline Tests/Admin/AdminHelperTest.php]: If you as the writer of the test decide to make it `covered`, then add the annotation.

but `@covers a` is wrong in this case, but would not result in any coverage.

The order is:
* execute code
* if executed code should be covered by annotation -> mark the executed lines covered, otherwise not.

Comment (greg0ire) [inline Tests/Admin/AdminHelperTest.php]: > but @covers a is wrong in this case, but would not result in any coverage of a() at all.

I updated my code to move the annotation to testA(), but I don't get your sentence…

Comment (OskarStark) [inline Tests/Admin/AdminHelperTest.php]: `testA` is not calling `a` or `c`

With **and** without `@covers`annotation, no coverage is done.

Let me explain it a little bit more in detail!
```php
<?php

class Foo {
    public function a ()
    {
       $this->b();
       $this->c();
    }
    public function b ()
    {
    }
    private function c ()
    {
    }
}
```

```php
class FooTest {
    /**
     *
     * @covers a
     * @covers c
     */
    public function testA()
    {
        $this->assert('test', $this->a());
    }
```

We got the following coverage:
* coverage for Foo::a()
* NO coverage for Foo::a(), because i didn't annotate it
* coverage for Foo::c()

But i would not set `@covers c`, because maybe i didn't test ALL the edge cases of `c()`, so this would be better:

```php
class FooTest {
    /**
     *
     * @covers a
     */
    public function testA()
    {
        $this->assert('test', $this->a());
    }
```

Now i am sure that i need to write a test for `b()` and `c()`, only callings some code lines is not a real test coverage for me and with the `@covers` annotation we can aschieve this more clear!

Comment (OskarStark) [inline Tests/Admin/AdminHelperTest.php]: @javierequiluz, sorry for bothering you, but can you may help us with this? maybe you already discussed this for the symfony project?

Comment (javiereguiluz) [inline Tests/Admin/AdminHelperTest.php]: @OskarStark basically we changed our position about this a few months ago (I cannot find the link to the exact discussion). We no longer care about `@covers` and `@codeCoverageIgnore` because in our opinion is not worth it to maintain them. After all, we don't follow TDD or want to achieve 100% coverage, so it's not that important for us.

Comment (greg0ire) [inline Tests/Admin/AdminHelperTest.php]: > After all, we don't follow TDD or want to achieve 100% coverage, so it's not that important for us.

Maybe we should start increasing code coverage before worrying about this ?

Comment (OskarStark) [inline Tests/Admin/AdminHelperTest.php]: > Maybe we should start increasing code coverage before worrying about this ?

exactly this is why i would implement this

> @OskarStark basically we changed our position about this a few months ago (I cannot find the link to the exact discussion). We no longer care about @covers and @codeCoverageIgnore because in our opinion is not worth it to maintain them. After all, we don't follow TDD or want to achieve 100% coverage, so it's not that important for us.

Thanks for your feedback on this 👍

Comment (Soullivaneuh) [inline Tests/Admin/AdminHelperTest.php]: Totally agree with @javiereguiluz for my case.
